src/compiler/checker.ts

```diff
@@ -32208,11 +32209,35 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return undefined;
     }
 
+    function getContextualPaddingType(node: BindingPattern | BindingElement): Type | undefined {
+        const index = findContextualNode(node, /*includeCaches*/ true);
+        if (index >= 0) {
+            return contextualTypes[index];
+        }
+        const { parent } = node;
+        switch (parent.kind) {
+            case SyntaxKind.ObjectBindingPattern: {
+                const type = getContextualPaddingType(parent);
+                const name = type && getPropertyNameFromBindingElement(node as BindingElement);
+                return name ? getTypeOfPropertyOfType(type, name) : undefined;
+            }
+            case SyntaxKind.BindingElement: {
+                if (parent.parent.kind === SyntaxKind.ArrayBindingPattern) {
+                    break;
+                }
+                const type = getContextualPaddingType(parent.parent);
+                const name = type && getPropertyNameFromBindingElement(parent);
+                return name ? getTypeOfPropertyOfType(type, name) : undefined;
+            }
+        }
+        return undefined;
+    }
+
     function pushCachedContextualType(node: Expression) {
         pushContextualType(node, getContextualType(node, /*contextFlags*/ undefined), /*isCache*/ true);
     }
 
-    function pushContextualType(node: Expression, type: Type | undefined, isCache: boolean) {
+    function pushContextualType(node: Node, type: Type | undefined, isCache: boolean) {

```

Comment:
Can this be a narrower type than just Node?

Reply:
With this implementation, it could be `Expression | BindingPattern | BindingElement` if you'd prefer that.

2 extra notes that come to my mind now:
- maybe I shouldn't be reusing the same `contextualTypes` stack for this and I should just introduce a new one? It seemed fine here to avoid creating a new but then reusing `getContextualType` wasn't possible so it's slightly weird now that I'm only using part of the existing pattern . Perhaps with a new context flag it would be possible but introducing `getContextualPaddingType` seemed easier
- if this gets approved, it might be a good idea to explore expanding `getContextualPaddingType` with support for array patterns and using `pushContextualType`/`popContextualType` in `padTupleType`. Since this PR tries to address a regression in a backportable manner I refrained from doing this to avoid any new unforeseen behavioral changes